### PR TITLE
Add `DeitaFiltering` step

### PR DESCRIPTION
**Pull request by gabrielmbmb**
## Description

This PR adds the `DeitaFiltering` steps and introduces some parameters in `EvolQuality` and `EvolInstruct` that allows including the original instruction in the list of `evolved_{response,instruction}`. In addition, it adds the `ExpandColumns` and `GenerateConversation` steps auxiliary steps that were needed to build the whole DEITA pipeline.

## Pipeline

```python
from distilabel.llm.huggingface.transformers import TransformersLLM
from distilabel.llm.openai import OpenAILLM
from distilabel.pipeline.local import Pipeline
from distilabel.steps.conversation import ConversationTemplate
from distilabel.steps.deita import DeitaFiltering
from distilabel.steps.expand import ExpandColumns
from distilabel.steps.generators.huggingface import LoadHubDataset
from distilabel.steps.task.complexity_scorer import ComplexityScorer
from distilabel.steps.task.evol_instruct.base import EvolInstruct
from distilabel.steps.task.evol_quality.base import EvolQuality
from distilabel.steps.task.generate_embeddings import GenerateEmbeddings
from distilabel.steps.task.quality_scorer import QualityScorer
from distilabel.steps.globals.huggingface import PushToHub


with Pipeline(name="DEITA", description="") as pipeline:
    load_data = LoadHubDataset(
        name="load_data", batch_size=100, output_mappings={"prompt": "instruction"}
    )

    evol_instruction_complexity = EvolInstruct(
        name="evol_instruction_complexity",
        llm=OpenAILLM(model="gpt-3.5-turbo"),
        num_evolutions=5,
        store_evolutions=True,
        generate_answers=True,
        include_original_instruction=True,
    )

    instruction_complexity_scorer = ComplexityScorer(
        name="instruction_complexity_scorer",
        llm=OpenAILLM(model="gpt-3.5-turbo"),
        input_mappings={"instructions": "evolved_instructions"},
    )

    expand_evolved_instructions = ExpandColumns(
        name="expand_evolved_instructions",
        columns=["evolved_instructions", "answers", "scores"],
        output_mappings={
            "evolved_instructions": "evolved_instruction",
            "answers": "answer",
            "scores": "evol_instruction_score",
        },
    )

    evol_response_quality = EvolQuality(
        name="evol_response_quality",
        llm=OpenAILLM(model="gpt-3.5-turbo"),
        num_evolutions=5,
        store_evolutions=True,
        include_original_response=True,
        input_mappings={
            "instruction": "evolved_instruction",
            "response": "answer",
        },
    )

    response_quality_scorer = QualityScorer(
        name="response_quality_scorer",
        llm=OpenAILLM(model="gpt-3.5-turbo"),
        input_mappings={
            "instruction": "evolved_instruction",
            "responses": "evolved_responses",
        },
    )

    expand_evolved_responses = ExpandColumns(
        name="expand_evolved_responses",
        columns=["evolved_responses", "scores"],
        output_mappings={
            "evolved_responses": "evolved_response",
            "scores": "evol_response_score",
        },
    )

    generate_conversation = ConversationTemplate(
        name="generate_conversation",
        input_mappings={
            "instruction": "evolved_instruction",
            "response": "evolved_response",
        },
    )

    push_to_hub_after_conversation = PushToHub(
        name="push_to_hub_after_conversation",
    )

    generate_embeddings = GenerateEmbeddings(
        name="generate_embeddings",
        llm=TransformersLLM(
            model="argilla/notus-7b-v1",
            device="cuda",
            torch_dtype="float16",
        ),
        input_mappings={"text": "conversation"},
        input_batch_size=5,
    )

    push_to_hub_after_embeddings = PushToHub(
        name="push_to_hub_after_embeddings",
    )

    deita_filtering = DeitaFiltering(name="deita_filtering")

    load_data.connect(evol_instruction_complexity)
    evol_instruction_complexity.connect(instruction_complexity_scorer)
    instruction_complexity_scorer.connect(expand_evolved_instructions)
    expand_evolved_instructions.connect(evol_response_quality)
    evol_response_quality.connect(response_quality_scorer)
    response_quality_scorer.connect(expand_evolved_responses)
    expand_evolved_responses.connect(generate_conversation)
    generate_conversation.connect(generate_embeddings)
    generate_conversation.connect(push_to_hub_after_conversation)
    generate_embeddings.connect(deita_filtering)
    generate_embeddings.connect(push_to_hub_after_embeddings)

if __name__ == "__main__":
    distiset = pipeline.run(
        parameters={
            "load_data": {
                "repo_id": "HuggingFaceH4/instruction-dataset",
                "split": "test",
            },
            "evol_instruction_complexity": {
                "llm": {
                    "generation_kwargs": {"max_new_tokens": 512, "temperature": 0.7}
                }
            },
            "instruction_complexity_scorer": {
                "llm": {"generation_kwargs": {"temperature": 0.0}}
            },
            "evol_response_quality": {
                "llm": {
                    "generation_kwargs": {"max_new_tokens": 512, "temperature": 0.7}
                }
            },
            "response_quality_scorer": {
                "llm": {"generation_kwargs": {"temperature": 0.0}}
            },
            "push_to_hub_after_conversation": {
                "repo_id": "distilabel-internal-testing/deita-after-conversation",
            },
            "push_to_hub_after_embeddings": {
                "repo_id": "distilabel-internal-testing/deita-after-embeddings",
            },
            "deita_filtering": {"data_budget": 3000, "diversity_threshold": 0.04},
        },
        use_cache=True,
    )
    distiset.push_to_hub("distilabel-internal-testing/deita")
```

### CODE REVIEW

File: src/distilabel/steps/deita.py [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+    )
+
+    _NearestNeighbors: Type["NearestNeighbors"] = PrivateAttr(...)
+
+    def load(self) -> None:
+        try:
+            from sklearn.neighbors import NearestNeighbors
+        except ImportError as ie:
+            raise ImportError(
+                "`scikit-learn` is not installed. Please install it using `pip install huggingface-hub`."
```
Comment (davidberenstein1957) [inline src/distilabel/steps/deita.py:84]: is this included in `huggingface-hub` ?

File: src/distilabel/steps/deita.py [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+        """Computes the DEITA score for each instruction-response pair. The DEITA score is
+        the product of the instruction score and the response score.
+
+        Args:
+            inputs: The input data.
+
+        Returns:
+            The input data with the DEITA score computed.
+        """
+        for input_ in inputs:
```
Comment (davidberenstein1957) [inline src/distilabel/steps/deita.py:107]: Not sure about returning the score if the computation was faulty. people might assume their response and instruction where good, while in practice only either one of them might have been correct.

File: src/distilabel/steps/deita.py [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+
+        Returns:
+            The input data with the cosine distance computed.
+        """
+        embeddings = [input["embedding"] for input in inputs]
+        if self.normalize_embeddings:
+            embeddings = self._normalize_embeddings(embeddings)
+        self._logger.info("📏 Computing nearest neighbor distance...")
+        nn = self._NearestNeighbors(
+            n_neighbors=2, metric="cosine", algorithm="brute"
```
Comment (davidberenstein1957) [inline src/distilabel/steps/deita.py:150]: is there a reason these are fixed?

File: src/distilabel/steps/deita.py [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+        ).fit(embeddings)
+        distances, _ = nn.kneighbors(embeddings, return_distance=True)
+        for distance, input in zip(distances, inputs):
+            input["nearest_neighbor_distance"] = distance[-1]
+        return inputs
+
+    def _normalize_embeddings(self, embeddings: List[np.ndarray]) -> List[np.ndarray]:
```
Comment (davidberenstein1957) [inline src/distilabel/steps/deita.py:157]: isnt't there a built-in function to do this within Sklearn?

File: src/distilabel/steps/deita.py (hunk 3 above)
Comment (gabrielmbmb) [reply src/distilabel/steps/deita.py:150]: yes, we just want the distance with respect to the first nearest neighbor. The first neighbor will always be the record itself, therefore we take the second neighbor which is the real one. For the algorithm, we don't want any approximation and it's fast, therefore `brute`. metric can be a `RuntimeParameter`.

File: src/distilabel/steps/deita.py (hunk 4 above)
Comment (gabrielmbmb) [reply src/distilabel/steps/deita.py]: yes, I removed completely scikit-learn dep, as we already have numpy as dep and they are easy functions to implement

File: src/distilabel/steps/deita.py (hunk 2 above)
Comment (gabrielmbmb) [reply src/distilabel/steps/deita.py:121]: I've added the `deita_score_computed_with` which indicates which scores were used to compute `deita_score`
